src/libtego/source/context.cpp

```diff
@@ -0,0 +1,1030 @@
+        vm["UseBridges"] = "1";
+    }
+
+    this->torControl->setConfiguration(vm);
+}
+
+void tego_context::save_tor_daemon_config()
+{
+    TEGO_THROW_IF_NULL(this->torControl);
+    logger::trace();
+    this->torControl->saveConfiguration();
+}
+
+void tego_context::set_host_user_state(tego_host_user_state_t state)
+{
+    if (state == hostUserState)
+    {
+        return;
+    }
+
+    this->hostUserState = state;
+    this->callback_registry_.emit_host_user_state_changed(state);
+}
+
+std::unique_ptr<tego_user_id_t> tego_context::get_host_user_id() const
+{
+    TEGO_THROW_IF_NULL(this->identityManager);
+    auto userIdentity = this->identityManager->identities().first();
+
+    auto hostname = userIdentity->hostname().toUtf8();
+    tego_v3_onion_service_id serviceId(hostname.data(), TEGO_V3_ONION_SERVICE_ID_LENGTH);
+
+    return std::make_unique<tego_user_id_t>(serviceId);
+}
+
+tego_host_user_state_t tego_context::get_host_user_state() const
+{
+    return this->hostUserState;
+}
+
+void tego_context::send_chat_request(
+    const tego_user_id_t* user,
+    const char* message,
+    size_t messageLength)
+{
+    TEGO_THROW_IF_NULL(this->identityManager);
+    auto userIdentity = this->identityManager->identities().first();
+    auto contactsManager = userIdentity->getContacts();
+
+    contactsManager->createContactRequest(
+        QString::fromStdString(fmt::format("ricochet:{}", user->serviceId.data)),
+        (messageLength == 0) ? QString() : QString::fromUtf8(message, messageLength));
+}
+
+void tego_context::acknowledge_chat_request(
+        const tego_user_id_t* user,
+        tego_chat_acknowledge_t response)
+{
+    TEGO_THROW_IF_NULL(user);
+
+    logger::println("ack chat request from {}", user->serviceId.data);
+    logger::println("response : {}", (int)response);
+
+    TEGO_THROW_IF_NULL(this->identityManager);
+    auto userIdentity = this->identityManager->identities().first();
+    auto contactsManager = userIdentity->getContacts();
+    auto incomingRequestManager = contactsManager->incomingRequestManager();
+
+    auto hostname = QString("%1.onion").arg(user->serviceId.data).toUtf8();
+
+    auto incomingContactRequest = incomingRequestManager->requestFromHostname(hostname);
+    TEGO_THROW_IF_NULL(incomingContactRequest);
+
+    switch(response)
+    {
+        case tego_chat_acknowledge_accept:
+            incomingContactRequest->accept(nullptr);
+            break;
+        case tego_chat_acknowledge_reject:
+            incomingContactRequest->reject();
+            break;
+        case tego_chat_acknowledge_block:
+            incomingContactRequest->reject();
+            incomingRequestManager->addRejectedHost(hostname);
+            break;
+    }
+}
+
+tego_message_id_t tego_context::send_message(
+    const tego_user_id_t* user,
+    const std::string& message)
+{
+    TEGO_THROW_IF_NULL(user);
+    TEGO_THROW_IF_FALSE(message.size() > 0)
+
+    auto contactUser = getContactUser(user);
```

Comment:
`getContactUser` can return nullptr, should there be a check here?

Reply:
yes